app/src/main/java/com/github/ravifrancesco/softwaredevproject/GPSTracker.java

```diff
@@ -116,80 +99,146 @@ public Location getLocation() {
             e.printStackTrace();
         }
 
-        return location;
     }
 
-    // stop the GPS in the app
-    public void stopUsingGPS(){
-        if(locationManager != null){
-            locationManager.removeUpdates(GPSTracker.this);
-        }
+    /**
+     * Method that checks which providers are enabled for requesting location
+     * It will update class variable isNetworkEnabled and isGPSEnabled accordingly
+     */
+    private void checkLocationManagerStatus() {
+        locationManager = (LocationManager) mContext.getSystemService(LOCATION_SERVICE);
+
+        // getting network status
+        isNetworkEnabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
+
+        // getting GPS status
+        isGPSEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
     }
 
-    public double getLatitude(){
-        if(location != null){
-            latitude = location.getLatitude();
+    /**
+     * This method handles the request to update the location.
+     * First it checks if the app has the proper permissions, then it will request a new location
+     * through the location manager
+     *
+     * @param selectedProvider  string indicating the chosen provider for requesting location
+     */
+    private void setLocation(String selectedProvider) {
+        //check the permission
+        if (ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+            ActivityCompat.requestPermissions((Activity) mContext, new String[]{android.Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION}, 101);
         }
+        // request location
+        locationManager.requestLocationUpdates(
+                selectedProvider,
+                MIN_TIME_BW_UPDATES,
+                MIN_DISTANCE_CHANGE_FOR_UPDATES, this);
+        // set new location
```

Comment:
`MIN_TIME_BW_UPDATES` -> the time should be given in ms. The comment next to the definition says seconds (x1000 needed)